app/routes/task_routes.py

```diff
@@ -1 +1,144 @@
+    response = {
+        "task": new_task.to_dict()
+    }
+    return response, 201
+
+
+@tasks_bp.get("")
+def get_all_tasks():
+    description_param = request.args.get("description")
+    title_param = request.args.get("title")
+    sort_param = request.args.get("sort")
+
+    query = db.select(Task)
+    if description_param:
+        query = query.where(Task.description.like(f"%{description_param}%"))
+
+    if title_param:
+        query = query.where(Task.title.like(f"%{title_param}"))
+    if sort_param == "asc":
+        query = query.order_by(asc(Task.title))
+    elif sort_param == "desc":
+        query = query.order_by(desc(Task.title))
+
+    tasks = db.session.scalars(query).all()
+    tasks_response = [task.to_dict() for task in tasks]
+    return tasks_response
+
+
+def validate_id(id):
+    if not id.isnumeric():
+        abort(make_response({"message": f"Tasks id {id} invalid"}, 400))
+
+
+@tasks_bp.get("/<task_id>")
+def get_one_task(task_id):
+    validate_id(task_id)
+    query = db.select(Task).where(Task.id == task_id)
+    task = db.session.scalar(query)
+
+    if not task:
+        response = {"message": f"{task_id} not found"}
+        abort(make_response(response, 404))
+
+    return {"task": task.to_dict()}
+
+
+@tasks_bp.put("/<task_id>")
+def update_task(task_id):
+    validate_id(task_id)
+    query = db.select(Task).where(Task.id == task_id)
+    task = db.session.scalar(query)
+
+    if task is None:
+        abort(make_response({"message": f"Task {task_id} not found"}, 404))
+
+    request_body = request.get_json()
+    task.title = request_body["title"]
+    task.description = request_body["description"]
+    if "is_complete" in request_body.keys() and request_body["is_complete"]:
+        task.completed_at = datetime.now()
+    else:
+        task.completed_at = None
+    db.session.commit()
+    return {
+        "task": task.to_dict()
+    }
+
+
+@tasks_bp.delete("/<task_id>")
+def delete_task(task_id):
+    validate_id(task_id)
+    query = db.select(Task).where(Task.id == task_id)
+    task = db.session.scalar(query)
+    if not task:
+        response = {"message": f"{task_id} not found"}
+        abort(make_response(response, 404))
+
+    db.session.delete(task)
+    db.session.commit()
+
+    return {
+        "details": f'Task {task_id} "{task.title}" successfully deleted'
+    }
+
+
+def change_task_status(task_id, is_completed):
+    task = Task.query.get(task_id)
+    if task is None:
+        response = {"message": f"Task {task_id} not found"}
+        abort(make_response(response, 404))
+
+    if is_completed:
+        completed_date = datetime.now()
+        task_text = f"Someone just completed the task '{task.title}'"
+    else:
+        completed_date = None
+        task_text = f"Someone just marked the task '{task.title}' incomplete"
+
+    notification_messages = SlackMessage("task-notifications", task_text)
+    notification_was_sent = slack_client.post_message(notification_messages)
```

Comment:
Recall that the project [requirement](https://github.com/Ada-C22/task-list-api/blob/main/ada-project-docs/wave_04.md#requirement-use-python-package-requests-to-make-http-calls) was to use the Python package `requests` to make 3rd party API calls. 

We want folks to get practice using `requests` and **not** use the `SlackClient` because not all 3rd party APIs wil provide you with a custom client.

@SalmaAnany Please let me know, as a repy to this comment, how you would re-write the code for calling Slack using `requests`.

Reply:
I used requests, please check the client, that's my own class using requests.

Reply:
Ah yes, I meant to remove this comment after I saw your SlackClient class! Thanks for pointing me there again.